Sources/Swift/Integrations/SessionReplay/SentryTouchTracker.swift

```diff
@@ -0,0 +1,146 @@
+        self.scale = CGAffineTransform(scaleX: CGFloat(scale), y: CGFloat(scale))
+    }
+    
+    func trackTouchFrom(event: UIEvent) {
+        guard let touches = event.allTouches else { return }
+        for touch in touches {
+            guard touch.phase == .began || touch.phase == .ended || touch.phase == .moved || touch.phase == .cancelled else { continue }
+            let info = trackedTouches[touch] ?? TouchInfo(id: touchId++)
+            let position = touch.location(in: nil).applying(scale)
+            let newEvent = TouchEvent(x: position.x, y: position.y, timestamp: event.timestamp, phase: touch.phase.toRRWebTouchPhase())
+            
+            switch touch.phase {
+            case .began:
+                info.startEvent = newEvent
+            case .ended, .cancelled:
+                info.endEvent = newEvent
+            case .moved:
+                // If the distance between two points is smaller than 10 points, we don't record the second movement.
+                // iOS event polling is fast and will capture any movement; we don't need this granularity for replay.
+                if let last = info.moveEvents.last, touchesDelta(last.point, position) < 10 { continue }
+                info.moveEvents.append(newEvent)
+                debounceEvents(in: info)
+            default:
+                continue
+            }
+            
+            trackedTouches[touch] = info
+        }
+    }
+    
+    private func touchesDelta(_ lastTouch: CGPoint, _ newTouch: CGPoint) -> CGFloat {
+        let dx = newTouch.x - lastTouch.x
+        let dy = newTouch.y - lastTouch.y
+        return sqrt(dx * dx + dy * dy)
+    }
+    
+    private func debounceEvents(in touchInfo: TouchInfo) {
+        guard touchInfo.moveEvents.count >= 3 else { return }
+        let subset = touchInfo.moveEvents.suffix(3)
+        if subset[subset.startIndex + 2].timestamp - subset[subset.startIndex + 1].timestamp > 0.5 {
+            // Don't debounce if the last two touches have at least a 500 millisecond difference to show this pause in the replay.
+            return
+        }
+        // If the last 3 touch points exist in a straight line, we don't need the middle point,
+        // because the representation in the replay with 2 or 3 points will be the same.
+        if arePointsCollinearSameDirection(subset[subset.startIndex].point, subset[subset.startIndex + 1].point, subset[subset.startIndex + 2].point) {
+            touchInfo.moveEvents.remove(at: touchInfo.moveEvents.count - 2)
+        }
+    }
+    
+    private func arePointsCollinearSameDirection(_ a: CGPoint, _ b: CGPoint, _ c: CGPoint) -> Bool {
+        var abAngle = atan2(b.x - a.x, b.y - a.y)
+        var bcAngle = atan2(c.x - b.x, c.y - b.y)
+
+        if abAngle * bcAngle < 0 { return false; }
+
+        abAngle += .pi
+        bcAngle += .pi
+
+        return abs(abAngle - bcAngle) < 0.05 || abs(abAngle - (2 * .pi - bcAngle)) < 0.05
+    }
```

Comment:
You can visually test the collinear tolerances in [here](https://jsfiddle.net/dhiogorb/8owgh1pb/3/)

Reply:
That's fancy. If the JS Fiddle is permanent it would be cool to add the link as a comment.